Source/items.cpp

```diff
@@ -3324,45 +3329,66 @@ void TryRandomUniqueItem(Item &item, _item_indexes idx, int8_t mLevel, int uper,
 
 	int32_t uidsIdx = std::max<int32_t>(0, GenerateRnd(static_cast<int32_t>(uids.size()))); // Index into uids, used to get a random uid from the uids vector.
 	int uid = uids[uidsIdx];                                                                // Actual unique id.
+	const UniqueItem &uniqueItem = UniqueItems[uid];
 
 	// If the selected unique was already generated, there is no need to fiddle with its parameters.
 	if (item._iUid == uid) {
-		UniqueItemFlags[uid] = true;
+		if (!gbIsMultiplayer) {
+			UniqueItemFlags[uid] = true;
+		}
 		return;
 	}
 
-	const UniqueItem &uniqueItem = UniqueItems[uid];
-	int targetLvl = 1; // Target level for reverse compatibility, since vanilla always takes the last applicable uid in the list.
-
-	// Set target level. Ideally we use uper 15 to have a 16% chance of generating a unique item.
-	if (uniqueItem.UIMinLvl - 4 > 0) { // Negative level will underflow. Lvl 0 items may have unintended consequences.
-		uper = 15;
-		targetLvl = uniqueItem.UIMinLvl - 4;
-	} else {
-		uper = 1;
-		targetLvl = uniqueItem.UIMinLvl;
+	// Find our own id to calculate the offset in validUniques and check if we can generate it reverse compaitble.
+	int uidOffset = -1;
+	bool uniqueWithSameMinLevelExistsAfterDesiredUnique = false;
+	for (size_t i = 0; i < validUniques.size(); i++) {
+		if (validUniques[i] == uid) {
+			// Vanilla always picks the last unique, so the offset is calculated from the back of the valid unique list.
+			uidOffset = static_cast<int>(validUniques.size() - i - 1);
+		} else if (uidOffset != -1 && UniqueItems[validUniques[i]].UIMinLvl == uniqueItem.UIMinLvl) {

```

Comment:
This was basically the cause of the infinite loops in the first place.
```suggestion
		} else if (uidOffset != -1 && UniqueItems[validUniques[i]].UIMinLvl <= uniqueItem.UIMinLvl) {
```

Reply:
Maybe I looked at the original code too much to reintroduce it? 😁 
Thanks for the highlight 🙂. 

Interestingly, I had a test debug function that checks if all uniques can be generated. This function was broken and so I missed this in my local testing. To avoid this in the future, I added a unit test that checks if all uniques can be generated. To make this work, I had to make some functions in items.h public (and out of the anonymous namespace). I did this in a separate commit, so hopefully it's easier to review.